### DISCUSSION TITLE
What is the best way to do something when a page is loaded?

### DISCUSSION THREAD

**Discussion by matt-grain** [reflex-dev/reflex#137] [Q&A] [answered]
Hi,

In case some stuff has to be initialized when a page is loaded, what is the best way to trigger this part of the code?
Currenty I'm using on_click on the link to the page but doesn't support page refresh obvously...

````python
pc.link(
    pc.image(src="/logo.jpg", width="auto", height="3em"),
    href="/page2",
    style=index_image,
    on_click=lambda: AppState.set_name("my name"),
),
````

**Comment by Alek99** [comment, marked as answer]
I think the best would be to use a computed var

**Reply by matt-grain** [reply]
and how would you call it?

**Reply by picklelo** [reply]
If you have a computed var in your state, you should be able to use it like any other var and it should hold the computed value of the function. But currently, the state doesn't have access to the current page, which is a limitation - I created  [a ticket](https://github.com/pynecone-io/pynecone/issues/138) for this.

There's a solution to your problem using [middleware](https://pynecone.io/docs/advanced-guide/middleware) but we are severely lacking in documentation for it currently.

```
class SetNameMiddleware(pc.Middleware):
    """Middleware to set the name when the page changes."""

    def preprocess(self, app, state, event):
        # (This is currently undocumented - the `hydrate` event is called on each page load)
        if event.name == pc.utils.get_hydrate_event(state):
            state.name = "my name"
```
then right before you compile your app, add the middleware:
```
app.add_middleware(SetNameMiddleware(), index=0)
```

The middleware can preprocess any events, so you can insert custom logic within there. I know it's probably a bit confusing, we will improve the docs for this.

**Reply by FHU-yezi** [reply]
So can we add callbacks to this event with `add_load_callback()` or something like this? I built a page to fetch MongoDB and display somthing, it has a input box and use datatable for data display.

When I refresh the page, it seems like the state has auto recovered, I got a empty input box but datatable is shown, with data before.

I use `data` state var to storage data, and add `state.set_data([])` at the top of my page func will broke the display function, it will display nothing when I input and press query button.

**Reply by picklelo** [reply]
@FHU-yezi since the state is stored in the backend, it will always persist through refreshes and page changes, until you close the browser tab. So that's why the datatable state didn't clear, while the input cleared since it's not attached to a state var.

We can definitely make this easier to add a callback on page loads. Where do you guys think would be the most natural place to put  this? Maybe in the `add_page` method?

```
app.add_page(index, on_load=State.on_page_load)
```

**Reply by FHU-yezi** [reply]
Yes, add it in add_page func will be a good choice.

BTW, I attached a state var to input value, and bind on_blur event for it, why it will be autoclear when page refresh?

**Reply by picklelo** [reply]
Oh, if you set the `value` prop it shouldn't autoclear - I only meant if it was an uncontrolled input, where you only set the `on_blur` without setting `value`.

**Reply by FHU-yezi** [reply]
> Oh, if you set the `value` prop it shouldn't autoclear - I only meant if it was an uncontrolled input, where you only set the `on_blur` without setting `value`.

I change state var in on_blur, because if I bind a state var to input box, it will use on_change event callback, which send request to backend when I type a char in it.

**Reply by FHU-yezi** [reply]
And for a chance, can we add a auto complete function to input box? Like datalist in HTML5, but some mobile browser has no support to this, so we should use a 3rd-party package, or use div to slove this.

**Reply by matt-grain** [reply]
> @FHU-yezi since the state is stored in the backend, it will always persist through refreshes and page changes, until you close the browser tab. So that's why the datatable state didn't clear, while the input cleared since it's not attached to a state var.
> 
> We can definitely make this easier to add a callback on page loads. Where do you guys think would be the most natural place to put this? Maybe in the `add_page` method?
> 
> ```
> app.add_page(index, on_load=State.on_page_load)
> ```

add_page would be called one time at start of the backend right?

**Reply by picklelo** [reply]
I was thinking it would be called each time the page loads, would that work? I created an [issue here ](https://github.com/pynecone-io/pynecone/issues/142 )to track this work.

**Reply by matt-grain** [reply]
yeah exactly! I give my vote :)

**Reply by heumsi** [reply]
I also give my vote :)

**Comment by Alek99** [comment]
Big thanks to @Lendemor, made a pr today to add this see here: https://github.com/pynecone-io/pynecone/commit/52aaf365c906e2522b079219b8af182b97efb6a3. This will be included in the next release.
